sway/input/cursor.c

```diff
@@ -37,162 +37,84 @@ static uint32_t get_current_time_msec(void) {
  * location, it is stored in **surface (it may not be a view).
  */
 struct sway_node *node_at_coords(
 		struct sway_seat *seat, double lx, double ly,
-		struct wlr_surface **surface, double *sx, double *sy) {
-	// find the output the cursor is on
-	struct wlr_output *wlr_output = wlr_output_layout_output_at(
-			root->output_layout, lx, ly);
-	if (wlr_output == NULL) {
-		return NULL;
-	}
-	struct sway_output *output = wlr_output->data;
-	if (!output || !output->enabled) {
-		// output is being destroyed or is being enabled
-		return NULL;
-	}
-	double ox = lx, oy = ly;
-	wlr_output_layout_output_coords(root->output_layout, wlr_output, &ox, &oy);
+		struct wlr_surface **_surface, double *sx, double *sy) {
 
-	// layer surfaces on the overlay layer are rendered on top
-	if ((*surface = layer_surface_at(output,
-				&output->layers[ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY],
-				ox, oy, sx, sy))) {
-		return NULL;
-	}
+	struct wlr_scene_node *scene_node =
+		wlr_scene_node_at(root->interactive, lx, ly, sx, sy);
 
-	// check for unmanaged views
-#if HAVE_XWAYLAND
-	struct wl_list *unmanaged = &root->xwayland_unmanaged;
-	struct sway_xwayland_unmanaged *unmanaged_surface;
-	wl_list_for_each_reverse(unmanaged_surface, unmanaged, link) {
-		struct wlr_xwayland_surface *xsurface =
-			unmanaged_surface->wlr_xwayland_surface;
-
-		double _sx = lx - unmanaged_surface->lx;
-		double _sy = ly - unmanaged_surface->ly;
-		if (wlr_surface_point_accepts_input(xsurface->surface, _sx, _sy)) {
-			*surface = xsurface->surface;
-			*sx = _sx;
-			*sy = _sy;
-			return NULL;
+	if (scene_node) {
+		// determine what wlr_surface we clicked on
+		if (scene_node->type == WLR_SCENE_NODE_SURFACE) {
+			*_surface = wlr_scene_surface_from_node(scene_node)->surface;
 		}
-	}
-#endif
 
-	if (root->fullscreen_global) {
-		// Try fullscreen container
-		struct sway_container *con = tiling_container_at(
-				&root->fullscreen_global->node, lx, ly, surface, sx, sy);
-		if (con) {
-			return &con->node;
-		}
-		return NULL;
-	}
-
-	// find the focused workspace on the output for this seat
-	struct sway_workspace *ws = output_get_active_workspace(output);
-	if (!ws) {
-		return NULL;
-	}
-
-	if (ws->fullscreen) {
-		// Try transient containers
-		for (int i = 0; i < ws->floating->length; ++i) {
-			struct sway_container *floater = ws->floating->items[i];
-			if (container_is_transient_for(floater, ws->fullscreen)) {
-				struct sway_container *con = tiling_container_at(
-						&floater->node, lx, ly, surface, sx, sy);
-				if (con) {
-					return &con->node;
+		// determine what container we clicked on
+		struct sway_container *con = NULL;
+		struct wlr_scene_node *current = scene_node;
+		while (current) {
+			if (current->data) {
+				struct sway_scene_descriptor *desc = current->data;
+
+				if (desc->type == SWAY_SCENE_DESC_CONTAINER) {
+					con = desc->data;
+					break;
+				}else if (desc->type == SWAY_SCENE_DESC_LAYER_SHELL) {
```

Comment:
Nit: spaces around the `else`, here and in a couple of other places in this PR (won't tag all of them to reduce noise).